Migrate Snowflake integration guide (#1258)
astronomer/docs: tutorials/airflow-snowflake.md

---
title: "Orchestrating Snowflake Queries with Airflow"
description: "How to use Airflow to get enhanced observability and compute savings while orchestrating your Snowflake jobs."
id: airflow-snowflake
sidebar_label: Snowflake
---

> ## Overview

Reviewer: ```suggestion

```

> [Snowflake](https://www.snowflake.com/) is one of the most commonly used data warehouses and features heavily in data engineering, data science, and data analytics workflows. Orchestrating Snowflake queries as part of a data pipeline is one of the most common use cases for Airflow. Using Airflow with Snowflake is straightforward, and there are multiple open source packages, tools, and integrations that can help you take your Snowflake query orchestration to the next level.

Reviewer: ```suggestion
[Snowflake](https://www.snowflake.com/) is one of the most commonly used data warehouses. Orchestrating Snowflake queries as part of a data pipeline is one of the most common Airflow use cases. Using Airflow with Snowflake is straightforward, and there are multiple open source packages, tools, and integrations that can help you realize the full potential of your existing Snowflake instance.
```

> This guide covers everything you need to know to make the most out of Airflow and Snowflake, including:

Reviewer: ```suggestion
This guide covers the following topics:
```

> - Using Snowflake providers, including what modules are out there and how to implement deferrable versions of common operators.

Reviewer: ```suggestion
- Using Snowflake providers and available modules, and how to implement deferrable versions of common operators.
```

- Leveraging the OpenLineage Airflow integration to get data lineage and enhanced observability from your Snowflake jobs.
- Using the Astro SDK for the next generation of DAG authoring for Snowflake query tasks.
- General best practices and considerations when interacting with Snowflake from Airflow.

> ## Using Snowflake Providers

Reviewer: ```suggestion
## Using Snowflake providers
```

> There are multiple open source packages that you can use to orchestrate Snowflake in Airflow:

Reviewer: ```suggestion
The following are some of the open source packages that you can use to orchestrate Snowflake in Airflow:
```

> - The [Snowflake provider package](https://registry.astronomer.io/providers/snowflake) contains hooks, operators, and transfer operators for Snowflake maintained by the Airflow community.

Reviewer: ```suggestion
- The [Snowflake provider package](https://registry.astronomer.io/providers/snowflake) is maintained by the Airflow community and contains hooks, operators, and transfer operators for Snowflake.
```

- The [Astronomer Providers](https://github.com/astronomer/astronomer-providers) package contains deferrable operators built and maintained by Astronomer, including a deferrable version of the `SnowflakeOperator`.

> - The [Common SQL provider package](https://registry.astronomer.io/providers/common-sql) contains SQL check operators that can be used to perform data quality checks against data in Snowflake.

Reviewer: ```suggestion
- The [Common SQL provider package](https://registry.astronomer.io/providers/common-sql) contains SQL check operators that can be used to perform data quality checks against Snowflake data.
```

> To leverage all of available Snowflake modules, install all three packages in your Airflow environment. If you use the Astro CLI, you can do this by adding the following three lines to your `requirements.txt` file of your Astro project:

Reviewer: ```suggestion
To leverage all of the available Snowflake modules, install all three packages in your Airflow environment. If you use the Astro CLI, add the following three lines to your Astro project `requirements.txt` file:
```

```bash
apache-airflow-providers-snowflake
apache-airflow-providers-common-sql
astronomer-providers[snowflake]
```

> Modules for orchestrating basic queries and functions in Snowflake include:

Reviewer: ```suggestion
The following modules are available for orchestrating basic queries and functions in Snowflake:
```

- [`SnowflakeOperator`](https://registry.astronomer.io/providers/snowflake/modules/snowflakeoperator): Executes a SQL query in Snowflake. This operator is part of `apache-airflow-providers-snowflake`.
- [`S3ToSnowflakeOperator`](https://registry.astronomer.io/providers/snowflake/modules/s3tosnowflakeoperator): Executes a COPY command to transfer data from S3 into Snowflake.
- [`SnowflakeToSlackOperator`](https://registry.astronomer.io/providers/snowflake/modules/snowflaketoslackoperator): Executes a SQL query in Snowflake and sends the results to Slack.
- `SnowflakeOperatorAsync`: The [deferrable](https://www.astronomer.io/guides/deferrable-operators) version of the `SnowflakeOperator`, executes a SQL query in Snowflake.
- [`SnowflakeHook`](https://registry.astronomer.io/providers/snowflake/modules/snowflakehook): A hook abstracting the Snowflake API. Generally, you only need to use this hook when creating a custom operator or function. This hook is part of `apache-airflow-providers-snowflake`.
- `SnowflakeHookAsync`: The [deferrable](https://www.astronomer.io/guides/deferrable-operators) version of the `SnowflakeHook`, abstracts the Snowflake API.

> Modules for orchestrating data quality checks in Snowflake include:

Reviewer: ```suggestion
The following are the available modules for orchestrating data quality checks in Snowflake:
```

- [`SQLColumnCheckOperator`](https://registry.astronomer.io/providers/common-sql/modules/sqlcolumncheckoperator): Performs a data quality check against columns of a given table. Using this operator with Snowflake requires a Snowflake connection ID, the name of the table to run checks on, and a `column_mapping` describing the relationship between columns and tests to run.
- [`SQLTableCheckOperator`](https://registry.astronomer.io/providers/common-sql/modules/sqltablecheckoperator): Performs a data quality check against a given table. Using this operator with Snowflake requires a Snowflake connection ID, the name of the table to run checks on, and a checks dictionary describing the relationship between the table and the tests to run.

> Note that the `apache-airflow-providers-snowflake` package also contains operators that can be used to run data quality checks in Snowflake, including the `SnowflakeCheckOperator`, `SnowflakeValueCheckOperator`, and `SnowflakeIntervalCheckOperator`. However, these operators are not as flexible as the operators in `apache-airflow-providers-common-sql` and will likely be deprecated in a future version of Airflow. Astronomer recommends using `apache-airflow-providers-common-sql` for the most up to date data quality check operators. For more details on using the SQL check operators, check out [this guide](https://www.astronomer.io/guides/airflow-sql-data-quality-tutorial).

Reviewer: ```suggestion
Although the `apache-airflow-providers-snowflake` package contains operators that you can use to run data quality checks in Snowflake, Astronomer recommends that you use `apache-airflow-providers-common-sql` instead. For more information about using the SQL check operators, see [Airflow Data Quality Checks with SQL Operators](https://www.astronomer.io/guides/airflow-sql-data-quality-tutorial).
```

Let's walk through an example of how to use some of these modules in a DAG that implements a write, audit, publish pattern with data quality checks.

### Example Implementation

:::info 

Example code using various Snowflake operators can be found on the [Astronomer Registry](https://registry.astronomer.io/dags/complex-snowflake-transform).

:::

> The example DAG below runs a write, audit, publish pattern to showcase loading and data quality checking with Snowflake. The following steps are completed:

Reviewer: ```suggestion
The following example DAG runs a write, audit, and publish pattern to showcase loading and quality checking Snowflake data. The DAG completes the following steps:
```

> - Simultaneously create tables in Snowflake for the production data and the raw data that needs to be audited using the `SnowflakeOperator`. Note that these tasks are not implemented with the deferrable version of the `SnowflakeOperator` because `CREATE TABLE` statements typically run very quickly.

Reviewer: ```suggestion
- Simultaneously creates tables in Snowflake for the production data and the raw data that needs to be audited using the `SnowflakeOperator`. These tasks are not implemented with the deferrable version of the `SnowflakeOperator` because `CREATE TABLE` statements typically run very quickly.
```

> - Load data into the audit table using the `SnowflakeOperatorAsync`. This task *is* deferred to save on compute, because loading data can take some time if the dataset is large. Note that to use the deferrable operator, you must have a [triggerer running](https://www.astronomer.io/guides/deferrable-operators#running-deferrable-tasks-in-your-airflow-environment) in your Airflow environment.

Reviewer: ```suggestion
- Loads data into the audit table using the `SnowflakeOperatorAsync`. This task is deferred to save on computational resources, because loading data can take some time if the dataset is large. To use the deferrable operator, you must have a [triggerer running](https://www.astronomer.io/guides/deferrable-operators#running-deferrable-tasks-in-your-airflow-environment) in your Airflow environment.
```

> - Run data quality checks on the audit table to ensure that no erroneous data is moved to production. This task group includes column checks using the `SQLColumnCheckOperator` and table checks using the `SQLTableCheckOperator`.

Reviewer: ```suggestion
- Runs data quality checks on the audit table to ensure that no erroneous data is moved to production. This task group includes column checks using the `SQLColumnCheckOperator` and table checks using the `SQLTableCheckOperator`.
```

> - If the data quality checks are passed, copy data from the audit table into the production table using the `SnowflakeOperatorAsync`.

Reviewer: ```suggestion
- If the data quality checks are passed, copies data from the audit table into the production table using the `SnowflakeOperatorAsync`.
```

> - Delete the audit table since it only contained temporary data.

Reviewer: ```suggestion
- Deletes the audit table because it only contains temporary data.
```

All of these tasks rely on parameterized SQL scripts that are stored in the `include/sql/` directory and can be found on the [Astronomer Registry](https://registry.astronomer.io/dags/complex-snowflake-transform).

> The DAG looks like this:

Reviewer: ```suggestion

```

        col_string = snowflake_load_column_string(table_props)

        copy_snowflake_audit_to_production_table = SnowflakeOperator(
            task_id="copy_snowflake_audit_to_production_table",
            sql="copy_forestfire_snowflake_audit.sql",
            params={
                "table_name": SNOWFLAKE_FORESTFIRE_TABLE,
                "audit_table_name": f"{SNOWFLAKE_FORESTFIRE_TABLE}_AUDIT",
                "table_schema": table_props,
                "col_string": col_string,
            },
            trigger_rule="all_success"
        )

    delete_audit_table = SnowflakeOperator(
        task_id="delete_audit_table",
        sql="delete_forestfire_table.sql",
        params={"table_name": f"{SNOWFLAKE_FORESTFIRE_TABLE}_AUDIT"},
        trigger_rule="all_success"
    )

    begin = EmptyOperator(task_id="begin")
    end = EmptyOperator(task_id="end")

    chain(
        begin,
        [create_forestfire_production_table, create_forestfire_audit_table],
        load_data,
        quality_check_group,
        copy_snowflake_audit_to_production_table,
        delete_audit_table,
        end
    )
```

![Snowflake DAG Graph](/img/guides/snowflake_dag_graph.png)

> Note that to run this DAG, you need an Airflow connection to your Snowflake instance. This DAG uses a connection called `snowflake_default`. Your connection should have a type of `Snowflake` include the following information:

Reviewer: ```suggestion
To run this DAG, you need an Airflow connection to your Snowflake instance. This DAG uses a connection named `snowflake_default`. Your connection should have a type of `Snowflake` include the following information:
```

```yaml

> Host: <Your Snowflake host, e.g. `account.region.snowflakecomputing.com`>

Reviewer: ```suggestion
Host: <Your Snowflake host. For example, `account.region.snowflakecomputing.com`>
```

Reviewer: Should the format of user entries be <your-snowflake-host>? If so, update all user entries.

> Host: <Your Snowflake host, e.g. `account.region.snowflakecomputing.com`>
> Schema: <Your schema>
> Login: <Your login>
> Password: <Your password>
> Account: <Your Snowflake account>
> Database: <Your database>
> Region: <Your account region>
> Role: <Your role>
> Warehouse: <Your warehouse>

Author: ```suggestion
Host: <your-snowflake-host> # For example, `account.region.snowflakecomputing.com`
Schema: <your-schema>
Login: <your-login>
Password: <your-password>
Account: <your-snowflake-account>
Database: <your-database>
Region: <your-account-region>
Role: <your-role>
Warehouse: <your-warehouse>
```

```

## Enhanced Observability with OpenLineage

> The [OpenLineage project](https://openlineage.io/) maintains an integration with Airflow that allows users to obtain and view lineage data from their Airflow tasks. As long as an extractor exists for the operator being used, lineage data will be generated automatically from each task instance. For introductory information on how OpenLineage works with Airflow, check out [this guide](https://www.astronomer.io/guides/airflow-openlineage).

Reviewer: ```suggestion
The [OpenLineage project](https://openlineage.io/) integration with Airflow lets you obtain and view lineage data from your  Airflow tasks. As long as an extractor exists for the operator being used, lineage data is generated automatically from each task instance. For an overview of how OpenLineage works with Airflow, see [OpenLineage and Airflow](https://www.astronomer.io/guides/airflow-openlineage).
```

> Because `SnowflakeOperator` and `SnowflakeOperatorAsync` have an extractor, you can use lineage metadata to answer questions across DAGs such as:

Reviewer: ```suggestion
Because `SnowflakeOperator` and `SnowflakeOperatorAsync` have an extractor, you can use lineage metadata to answer the following questions across DAGs:
```

- How does data stored in Snowflake flow through my DAGs? Are there any upstream dependencies?
- What downstream data does a task failure impact?
- Where did a change in data format originate?

> At a high level, the interaction between OpenLineage,Airflow, and Snowflake works like this:

Reviewer: ```suggestion
This image shows an overview of the interaction between OpenLineage,Airflow and Snowflake:
```

![Snowflake Openlineage](/img/guides/snowflake_openlineage_architecture.png)

Note that to view lineage data from your DAGs you need to have OpenLineage installed in your Airflow environment and a lineage front end running. For [Astro users](https://docs.astronomer.io/astro/data-lineage), lineage is enabled automatically. For users working with open source tools, you can run Marquez locally and connect it to your Airflow environment following the instructions in [this guide](https://www.astronomer.io/guides/airflow-openlineage).

> To show an example of lineage resulting from Snowflake orchestration, we'll look at the write, audit, publish DAG from the example above. Note that screenshots below are from the Datakin UI integrated with Astro, but Marquez will show similar information.

Reviewer: ```suggestion
To show an example of lineage resulting from Snowflake orchestration, you'll look at the write, audit, publish DAG from the previous example. The following image shows the Datakin UI integrated with Astro, but Marquez will show similar information.
```

![Lineage Graph](/img/guides/lineage_graph.png)

Looking at the lineage graph, you can see the flow of data from the creation of the table, to the insertion of data, to the data quality checks. If any failures were to happen, particularly in the data quality checks, the lineage graph would quickly tell us which datasets were impacted. If your work on this dataset expanded into other DAGs in Airflow, you would see those connections here as well, all in one place.

## DAG Authoring with the Astro Python SDK

> The Astro Python SDK is an open source DAG authoring tool maintained by Astronomer that simplifies the data transformation process between different environments, so you can focus solely on writing execution logic without worrying about Airflow orchestration logic. Details like creating dataframes, storing intermediate results, passing context and data between tasks, and creating Airflow task dependencies are all managed automatically.

Reviewer: ```suggestion
The Astro Python SDK is an open source DAG authoring tool maintained by Astronomer that simplifies the data transformation process between different environments, so you can focus solely on writing execution logic without worrying about Airflow orchestration logic. Details such as creating dataframes, storing intermediate results, passing context and data between tasks, and creating Airflow task dependencies are all managed automatically.
```

> The Astro Python SDK supports Snowflake as a data warehouse and can be used to simplify ETL workflows with Snowflake. For example, the following DAG moves data from S3 into Snowflake, performs some data transformations, and loads the resulting data into a reporting table.

Reviewer: ```suggestion
The Astro Python SDK supports Snowflake as a data warehouse and can be used to simplify ETL workflows with Snowflake. For example, the following DAG moves data from Amazon S3 into Snowflake, performs some data transformations, and loads the resulting data into a reporting table.
```

    customers_table = Table(
        name=SNOWFLAKE_CUSTOMERS,
        conn_id=SNOWFLAKE_CONN_ID,
    )

    # filter the orders data and then join with the customer table
    joined_data = join_orders_customers(filter_orders(orders_data), customers_table)

    # merge the joined data into our reporting table, based on the order_id .
    # If there's a conflict in the customer_id or customer_name then use the ones from
    # the joined data
    reporting_table = aql.merge(
        target_table=Table(
            name=SNOWFLAKE_REPORTING,
            conn_id=SNOWFLAKE_CONN_ID,
        ),
        source_table=joined_data,
        target_conflict_columns=["order_id"],
        columns=["customer_id", "customer_name"],
        if_conflicts="update",
    )

    purchase_dates = transform_dataframe(reporting_table)
```

Using Astro SDK `aql` functions, you are able to seamlessly transition between SQL transformations (`filter_orders` and `join_orders_customers`) to Python dataframe transformations (`transform_dataframe`). All intermediary data created by each task is automatically stored in Snowflake and made available to downstream tasks.

> For more detailed instructions on running this example DAG, check out the [Astro SDK Getting Started doc](https://github.com/astronomer/astro-sdk/blob/main/docs/getting-started/GETTING_STARTED.md).

Reviewer: ```suggestion
For more detailed instructions on running this example DAG, see [Astro SDK Getting Started](https://github.com/astronomer/astro-sdk/blob/main/docs/getting-started/GETTING_STARTED.md).
```

## Best practices and considerations

> Below are some best practices and considerations to keep in mind when orchestrating Snowflake queries from Airflow, based on the many implementations we've seen:

Reviewer: ```suggestion
The following are some best practices and considerations to keep in mind when orchestrating Snowflake queries from Airflow:
```

> - Wherever possible, use the deferrable version of operators to save on compute. This will result in costs savings and greater scalability for your Airflow environment.

Reviewer: ```suggestion
-  To reduce costs and improve the scalability of your Airflow environment, use the deferrable version of operators.
```

> - Set your default Snowflake query specifications like Warehouse, Role, Schema, etc. in the Airflow connection. Then overwrite those parameters for specific tasks as necessary in your operator definitions. This is cleaner and easier to read than adding `USE Warehouse XYZ;` statements within your queries.

Reviewer: ```suggestion
- Set your default Snowflake query specifications such as Warehouse, Role, Schema, and so on in the Airflow connection. Then overwrite those parameters for specific tasks as necessary in your operator definitions. This is cleaner and easier to read than adding `USE Warehouse XYZ;` statements within your queries.
```

- Pay attention to which Snowflake compute resources your tasks are using, as overtaxing your assigned resources can cause slowdowns in your Airflow tasks. It is generally recommended to have different warehouses devoted to your different Airflow environments to ensure DAG development and testing does not interfere with DAGs running in production.

> - Make use of [Snowflake stages](https://docs.snowflake.com/en/sql-reference/sql/create-stage.html) when loading data from an external system using Airflow. Transfer operators like the `S3ToSnowflake` operator will require a Snowflake stage be set up. Stages generally make it much easier to repeatedly load data of a particular format.

Reviewer: ```suggestion
- Make use of [Snowflake stages](https://docs.snowflake.com/en/sql-reference/sql/create-stage.html) when loading data from an external system using Airflow. Transfer operators such as the `S3ToSnowflake` operator require a Snowflake stage be set up. Stages generally make it much easier to repeatedly load data in a specific format.
```
